example/particles/particles.cpp

```diff
@@ -0,0 +1,362 @@
+  Real minx_i = pmb->coords.x1f(ib.s);
+  Real minx_j = pmb->coords.x2f(jb.s);
+  Real minx_k = pmb->coords.x3f(kb.s);
+
+  auto &x = swarm->GetReal("x").Get();
+  auto &y = swarm->GetReal("y").Get();
+  auto &z = swarm->GetReal("z").Get();
+  auto &vx = swarm->GetReal("vx").Get();
+  auto &vy = swarm->GetReal("vy").Get();
+  auto &vz = swarm->GetReal("vz").Get();
+  auto &weight = swarm->GetReal("weight").Get();
+
+  pmb->par_for(
+      "CreateSomeParticles", 0, swarm->get_max_active_index(),
+      KOKKOS_LAMBDA(const int n) {
+        if (new_particles_mask(n)) {
+          auto rng_gen = rng_pool.get_state();
+
+          // Randomly sample in space in this meshblock
+          x(n) = minx_i + nx_i * dx_i * rng_gen.drand();
+          y(n) = minx_j + nx_j * dx_j * rng_gen.drand();
+          z(n) = minx_k + nx_k * dx_k * rng_gen.drand();
+
+          // Randomly sample direction on the unit sphere, fixing speed
+          Real theta = acos(2. * rng_gen.drand() - 1.);
+          Real phi = 2. * M_PI * rng_gen.drand();
+          vx(n) = v * sin(theta) * cos(phi);
+          vy(n) = v * sin(theta) * sin(phi);
+          vz(n) = v * cos(theta);
+
+          weight(n) = 1.0;
+
+          rng_pool.free_state(rng_gen);
+        }
+      });
+
+  return TaskStatus::complete;
+}
+
+TaskStatus TransportSwarm(MeshBlock *pmb, int stage, std::vector<std::string> &stage_name,
+                          Integrator *integrator) {
+  auto swarm = pmb->real_containers.GetSwarmContainer()->Get("my particles");
+
+  int max_active_index = swarm->get_max_active_index();
+
+  Real dt = integrator->dt;
+
+  auto &x = swarm->GetReal("x").Get();
+  auto &y = swarm->GetReal("y").Get();
+  auto &z = swarm->GetReal("z").Get();
+  auto &vx = swarm->GetReal("vx").Get();
+  auto &vy = swarm->GetReal("vy").Get();
+  auto &vz = swarm->GetReal("vz").Get();
+  auto &mask = swarm->GetMask().Get();
+
+  Real dx_i = pmb->coords.dx1f(pmb->cellbounds.is(IndexDomain::interior));
+  Real dx_j = pmb->coords.dx2f(pmb->cellbounds.js(IndexDomain::interior));
+  Real dx_k = pmb->coords.dx3f(pmb->cellbounds.ks(IndexDomain::interior));
+  Real dx_push = std::min<Real>(dx_i, std::min<Real>(dx_j, dx_k));
+
+  IndexRange ib = pmb->cellbounds.GetBoundsI(IndexDomain::interior);
+  IndexRange jb = pmb->cellbounds.GetBoundsJ(IndexDomain::interior);
+  IndexRange kb = pmb->cellbounds.GetBoundsK(IndexDomain::interior);
+
+  Real x_min = pmb->coords.x1f(ib.s);
+  Real y_min = pmb->coords.x2f(jb.s);
+  Real z_min = pmb->coords.x3f(kb.s);
+  Real x_max = pmb->coords.x1f(ib.e + 1);
+  Real y_max = pmb->coords.x2f(jb.e + 1);
+  Real z_max = pmb->coords.x3f(kb.e + 1);
+
+  ParArrayND<Real> t("time", max_active_index + 1);
+
+  // Simple particle push: push particles half a zone width until they have
+  // traveled one integrator timestep's worth of time
+  pmb->par_for(
+      "TransportSwarm", 0, max_active_index, KOKKOS_LAMBDA(const int n) {
+        if (mask(n)) {
```

Comment:
Hmm. One thing worth thinking about is the limit where `mask(n) = 0` for all but like one particle. The code is still the speed of the push for that one particle, probably. I don't think that's a problem, but it might be worth highlighting in documentation or a comment.

Reply:
This is something that would be cleaned up by calling `Defrag` -- if `mask(n) = true` for only one particle, that particle will be moved to index 0 and `max_active_index` will be 0. I've added a note about `Defrag` to the documentation.